extensions/sinks/kafka/kafka.go

```diff
@@ -0,0 +1,221 @@
+
+	SASL_NONE  = "none"
+	SASL_PLAIN = "plain"
+	SASL_SCRAM = "scram"
+)
+
+type sinkConf struct {
+	Brokers           string `json:"brokers"`
+	Topic             string `json:"topic"`
+	DeliveryGuarantee string `json:"deliveryGuarantee"`
+	SaslAuthType      string `json:"saslAuthType"`
+	SaslUserName      string `json:"saslUserName"`
+	SaslPassword      string `json:"saslPassword"`
+}
+
+func (m *kafkaSink) Configure(props map[string]interface{}) error {
+	c := &sinkConf{
+		Brokers:           "localhost:9092",
+		Topic:             "",
+		DeliveryGuarantee: AT_LEAST_ONCE,
+		SaslAuthType:      SASL_NONE,
+	}
+	if err := cast.MapToStruct(props, c); err != nil {
+		return err
+	}
+	if len(strings.Split(c.Brokers, ",")) == 0 {
+		return fmt.Errorf("brokers can not be empty")
+	}
+	if c.Topic == "" {
+		return fmt.Errorf("topic can not be empty")
+	}
+	if !(c.DeliveryGuarantee == AT_MOST_ONCE || c.DeliveryGuarantee == AT_LEAST_ONCE) {
+		return fmt.Errorf("deliveryGuarantee incorrect")
+	}
+	if !(c.SaslAuthType == SASL_NONE || c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) {
+		return fmt.Errorf("saslAuthType incorrect")
+	}
+	if (c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) && (c.SaslUserName == "" || c.SaslPassword == "") {
+		return fmt.Errorf("username and password can not be empty")
+	}
+
+	m.c = c
+	return nil
+}
+
+func (m *kafkaSink) Open(ctx api.StreamContext) error {
+	ctx.GetLogger().Debug("Opening kafka sink")
+	var maxAttempts int
+	var err error
+	var mechanism sasl.Mechanism
+	//delivery guarantee
+	switch m.c.DeliveryGuarantee {
+	case AT_MOST_ONCE:
+		maxAttempts = 0
+	case AT_LEAST_ONCE:
+		maxAttempts = 10
+	default:
+		maxAttempts = 10
+	}
+	//sasl authentication type
+	switch m.c.SaslAuthType {
+	case SASL_PLAIN:
+		mechanism = plain.Mechanism{
+			Username: m.c.SaslUserName,
+			Password: m.c.SaslPassword,
+		}
+	case SASL_SCRAM:
+		mechanism, err = scram.Mechanism(scram.SHA512, m.c.SaslUserName, m.c.SaslPassword)
+		if err != nil {
+			return err
+		}
+	default:
+		mechanism = nil
+	}
+	brokers := strings.Split(m.c.Brokers, ",")
+	w := &kafkago.Writer{
+		Addr:                   kafkago.TCP(brokers...),
+		Topic:                  m.c.Topic,
+		Balancer:               &kafkago.LeastBytes{},
+		Async:                  false,
+		AllowAutoTopicCreation: true,
+		MaxAttempts:            maxAttempts,
+		RequiredAcks:           -1,
+		Transport: &kafkago.Transport{
+			SASL: mechanism,
+		},
+		WriteTimeout: 10 * time.Second,
+		ReadTimeout:  10 * time.Second,
+	}
+	m.writer = w
+	return nil
+}
+
+func (m *kafkaSink) Collect(ctx api.StreamContext, item interface{}) error {
+	logger := ctx.GetLogger()
+	logger.Debugf("kafka sink receive %s", item)
+	var decodedBytes []byte
+	var err error
+	decodedBytes, _, err = ctx.TransformOutput(item)
+	if err != nil {
+		return fmt.Errorf("kafka sink transform data error: %v", err)
+	}
+
+	vctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
Is this timeout needed as the kafkaWriteWithBackoff already has an implicit timeout